### PR TITLE
Partial progress commitment, looking for comments! - Daniel Shi

### PR DESCRIPTION
**Pull request by zdanielshi**
Completed the list_tuples exercise
Screen Shot 2022-04-28 at 6 56 23 PM

Completed the dict_labels exercise
Screen Shot 2022-04-28 at 6 55 07 PM

Partially completed the list_dicts exercise (still need to complete the item to return an dictionary with the employment rate)
Screen Shot 2022-04-28 at 6 56 17 PM

Still need to complete the last exercise: dict_columns

### CODE REVIEW

File: ds-prep-python-data-series/list_tuples.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+def last_appearing_rate(keys, values):
+  last_data_point = values[-1]
+  dictionary = {
+    keys[0] : last_data_point[0],
+    keys[1] : last_data_point[1]
+  }
+  return dictionary
+
+print(last_appearing_rate(columns, unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:46]: 👍

File: ds-prep-python-data-series/list_tuples.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+# Get the first five series points, in the order they appear, as a list of
+# tuples.
+def first_five_series_points(series):
+  return series[0:4]
+
+print(first_five_series_points(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:54]: This only produces the first four series points.

File: ds-prep-python-data-series/list_tuples.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+def check_years_present(value, data_series):
+  list_of_years = []
+  for x in data_series:
+    (year_value, rate_value) = x
+    list_of_years.append(year_value)
+  return value in list_of_years
+
+print(check_years_present(2000, unemployment_rates))
+print(check_years_present(2010, unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:66]: 👍

File: ds-prep-python-data-series/list_tuples.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+# Get the unemployment rate for the most recent year.
+
+def ue_rate(data_series):
+  data_series.sort()
+  year_data = data_series[-1]
+  return year_data[1]
+
+print(ue_rate(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:76]: The `.sort()` method permanently modifies the order of the series' points. Please try to accomplish this task without that:

---

image

File: ds-prep-python-data-series/list_tuples.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+def unique_years(data_series):
+  list_of_years = []
+  for x in data_series:
+    (year_value, rate_value) = x
+    list_of_years.append(year_value)
+  unique_years = set(list_of_years)
+  return unique_years
+
+print(unique_years(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:89]: 👍

File: ds-prep-python-data-series/list_tuples.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+def largest_ue_rate(data_series):
+  list_of_rates = []
+  for x in data_series:
+    (year_value, rate_value) = x
+    list_of_rates.append(rate_value)
+  largest_ue_rate = max(list_of_rates)
+  return largest_ue_rate
+
+print(largest_ue_rate(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:109]: 👍

File: ds-prep-python-data-series/list_tuples.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    years_list.append(year_value) # append year values to list
+    ue_rates_list.append(rate_value) # append ue values to list
+  for y in ue_rates_list: # run a loop on the ue values
+    e_rate = 100 - y # calculate the e value
+    e_rates_list.append(e_rate) # append the e values to the e list
+  merged_list = [(years_list[i], e_rates_list[i]) for i in range(0, len(years_list))] # merge the lists
+  return merged_list # return merged list
+
+print(employment_rate(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:131]: Looks like this produces the correct output, though there's a much simpler way to accomplish the same task.

File: ds-prep-python-data-series/list_tuples.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+def high_unemployment_rates(data_series):
+  high_ue = [] # define a list for UE > 7
+  for x in data_series: ## run a loop of the rates
+    if x[1] > 7: ## if UE value is higher than 7
+      high_ue.append(x) ## append to list
+  return high_ue ## return list
+
+print(high_unemployment_rates(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:144]: 👍

File: ds-prep-python-data-series/list_tuples.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    (year_value, rate_value) = x # unpack the tuple
+    all_ue_rates.append(rate_value) # append the rates value to the rates list
+  for y in all_ue_rates: # run a loop for the rates list
+    if y in ue_rates_dict: # if value exists
+      ue_rates_dict[y] += 1 # increment existing key by 1
+    else:# if value is new
+      ue_rates_dict[y] = 1 # create new key, increment by 1
+  return ue_rates_dict
+
+print(unique_values_dict(unemployment_rates))
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_tuples.py:162]: 👍

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

File: ds-prep-python-data-series/list_dicts.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    {"unemployment_rate": 5, "year": 2008}
+]
+
+# Write your code below this line:
+# Get the last-appearing series point, as a dict with the column names as keys.
+def last_data_series(data_series):
+  return data_series[-1]
+
+print(last_data_series(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:40]: 👍

File: ds-prep-python-data-series/list_dicts.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+  years_list = [] ## define years list
+  for x in data_series: ## run a loop for all rows
+    rate_list.append(x['unemployment_rate']) ## append rates values to rates list
+    years_list.append(x['year']) ## append years values to years list
+  merged_list = [(years_list[i], rate_list[i]) for i in range(0,len(years_list))] ## merge lists together into a list of tuples
+  merged_list_first_five = merged_list[0:5]
+  return merged_list_first_five ## isolate for the first five values
+
+print(first_five_data_series(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:56]: 👍

File: ds-prep-python-data-series/list_dicts.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+def check_years_present(data_series, year_value):
+  years_list = []
+  for x in data_series: ## run a loop to isolate each row
+    years_list.append(x['year']) ## index the 'year' to return the year values
+  return year_value in years_list ## append the individual year values to list
+
+print(check_years_present(unemployment_rates, 2000)) ## check for 2000
+print(check_years_present(unemployment_rates, 2010)) ## check for 2000
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:68]: 👍

File: ds-prep-python-data-series/list_dicts.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+# Get the unemployment rate for the most recent year.
+def ue_rate_most_recent_year(data_series):
+  years_list = [] ## define a years_list
+  rates_list = [] ## define a rates list
+  for x in data_series: ## run a loop to isolate each row
+    years_list.append(x['year']) ## append years values to the years_list
+    rates_list.append(x['unemployment_rate']) ## append unemployment rate values to the rates list
+  highest_year = years_list.index(max(years_list)) ## return the index value of the highest year
+  highest_year_rate = rates_list[highest_year] ## apply that index value to the rates list
+  return highest_year_rate ## return the rate
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:79]: 👍

File: ds-prep-python-data-series/list_dicts.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+# Get all the unique years included in the series, as a set.
+def unique_years(data_series):
+  years_list = [] ## define a years list
+  for x in data_series: ## run a loop to isolate each row
+    years_list.append(x['year']) ## append years values to the years list
+  unique_years_set = set(years_list) ## convert the unique years list to a set
+  return unique_years_set ## return the set
+
+print(unique_years(unemployment_rates))
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:93]: 👍

File: ds-prep-python-data-series/list_dicts.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+  rates_list = [] ## define a rates list
+  for x in data_series: ## run a loop to isolate each row
+    years_list.append(x['year']) ## append years values to the years_list
+    rates_list.append(x['unemployment_rate']) ## append unemployment rate values to the rates list
+  merged_list = [(years_list[i], rates_list[i]) for i in range(0,len(years_list))] ## merge the lists together
+  merged_list.sort() ## sort the years column
+  return merged_list
+
+print(ordered_by_year(unemployment_rates))  ## print
+print("---")
```
Comment (TimelikeClosure) [inline ds-prep-python-data-series/list_dicts.py:107]: Almost. While the list it sorted by the year, the output should only contain the values, not the years:

---

image

File: ds-prep-python-data-series/list_tuples.py (hunk 4 above)
Comment (zdanielshi) [reply ds-prep-python-data-series/list_tuples.py:76]: Oh that's interesting. And I could see how that would *really* mess with production code. So I guess the thing to do here would be to copy the collection somehow and run the sort there.

**Pull request closed without merging**